File: app/src/lib/demo/handEval.ts
const RANKS = 13;

export interface HandRank {
  category: number;
  kickers: [number, number, number, number, number];
}

function cardRank(card: number): number {
  return card % 13;
}

function cardSuit(card: number): number {
  return Math.floor(card / 13);
}

function findStraightHigh(rankCounts: number[]): number | null {
  if (
    rankCounts[12] > 0 &&
    rankCounts[0] > 0 &&
    rankCounts[1] > 0 &&
    rankCounts[2] > 0 &&
    rankCounts[3] > 0
  ) {
    return 3;
  }
  for (let high = RANKS - 1; high >= 4; high--) {
    let ok = true;
    for (let i = 0; i < 5; i++) {
      if (rankCounts[high - 4 + i] === 0) {
        ok = false;
        break;
      }
    }
    if (ok) return high;
  }
  return null;
}

function evaluateFive(cards: number[]): HandRank {
  const ranks = cards.map(cardRank).sort((a, b) => b - a);
  const suits = cards.map(cardSuit);
  const rankCounts = new Array(RANKS).fill(0);
  const suitCounts = [0, 0, 0, 0];
  for (const r of ranks) rankCounts[r]++;
  for (const s of suits) suitCounts[s]++;

  const isFlush = suitCounts.some((c) => c >= 5);
  const straightHigh = findStraightHigh(rankCounts);

  if (isFlush && straightHigh !== null) {
    if (straightHigh === 12) return { category: 9, kickers: [12, 0, 0, 0, 0] };
    return { category: 8, kickers: [straightHigh, 0, 0, 0, 0] };
  }

  let quads: number | null = null;
  let trips: number | null = null;
  const pairs: number[] = [];
  for (let r = RANKS - 1; r >= 0; r--) {
    const count = rankCounts[r];
    if (count === 4) quads = r;
    else if (count === 3 && trips === null) trips = r;
    else if (count === 2 && pairs.length < 2) pairs.push(r);
  }

  if (quads !== null) {
    const kicker = [...Array(RANKS).keys()]
      .reverse()
      .find((r) => rankCounts[r] > 0 && r !== quads)!;
    return { category: 7, kickers: [quads, kicker, 0, 0, 0] };
  }

  if (trips !== null && pairs.length >= 1) {
    return { category: 6, kickers: [trips, pairs[0], 0, 0, 0] };
  }

  if (isFlush) {
    return { category: 5, kickers: [ranks[0], ranks[1], ranks[2], ranks[3], ranks[4]] };
  }

  if (straightHigh !== null) {
    return { category: 4, kickers: [straightHigh, 0, 0, 0, 0] };
  }

  if (trips !== null) {
    const kickers: number[] = [];
    for (let r = RANKS - 1; r >= 0 && kickers.length < 2; r--) {
      if (rankCounts[r] > 0 && r !== trips) kickers.push(r);
    }
    return { category: 3, kickers: [trips, kickers[0], kickers[1], 0, 0] };
  }

  if (pairs.length >= 2) {
    const highPair = Math.max(pairs[0], pairs[1]);
    const lowPair = Math.min(pairs[0], pairs[1]);
    const kicker = [...Array(RANKS).keys()]
      .reverse()
      .find((r) => rankCounts[r] > 0 && r !== highPair && r !== lowPair)!;
    return { category: 2, kickers: [highPair, lowPair, kicker, 0, 0] };
  }

  if (pairs.length === 1) {
    const kickers: number[] = [];
    for (let r = RANKS - 1; r >= 0 && kickers.length < 3; r--) {
      if (rankCounts[r] > 0 && r !== pairs[0]) kickers.push(r);
    }
    return { category: 1, kickers: [pairs[0], kickers[0], kickers[1], kickers[2], 0] };
  }

  return { category: 0, kickers: [ranks[0], ranks[1], ranks[2], ranks[3], ranks[4]] };
}

function compareRank(a: HandRank, b: HandRank): number {
  if (a.category !== b.category) return a.category - b.category;
  for (let i = 0; i < 5; i++) {
    if (a.kickers[i] !== b.kickers[i]) return a.kickers[i] - b.kickers[i];
  }
  return 0;
}

export function compareHands(a: HandRank, b: HandRank): number {
  if (a.category !== b.category) return a.category - b.category;
  for (let i = 0; i < 5; i++) {
    if (a.kickers[i] !== b.kickers[i]) return a.kickers[i] - b.kickers[i];
  }
  return 0;
}

export function evaluateHand(hole: [number, number], community: number[]): HandRank {
  const cards = [hole[0], hole[1], ...community.filter((c) => c < 52)];
  let best: HandRank = { category: 0, kickers: [0, 0, 0, 0, 0] };

  for (let i = 0; i < cards.length; i++) {
    for (let j = i + 1; j < cards.length; j++) {
      for (let k = j + 1; k < cards.length; k++) {
        for (let l = k + 1; l < cards.length; l++) {
          for (let m = l + 1; m < cards.length; m++) {
            const five = [cards[i], cards[j], cards[k], cards[l], cards[m]];
            const rank = evaluateFive(five);
            if (compareRank(rank, best) > 0) best = rank;
          }
        }
      }
    }
  }
  return best;
}

File: programs/solana_poker/src/hand_eval.rs
/// Texas Hold'em hand evaluation for 7 cards (2 hole + 5 community).
/// Cards are encoded 0-51: rank = card % 13 (0=2..12=A), suit = card / 13 (0-3).

const RANKS: usize = 13;
const SUITS: usize = 4;

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub struct HandRank {
    pub category: u8,
    pub kickers: [u8; 5],
}

impl HandRank {
    pub fn new(category: u8, kickers: [u8; 5]) -> Self {
        Self { category, kickers }
    }
}

pub fn evaluate_hand(hole: [u8; 2], community: &[u8]) -> HandRank {
    let mut cards = [0u8; 7];
    cards[0] = hole[0];
    cards[1] = hole[1];
    let mut n = 2usize;
    for &c in community {
        if c < 52 {
            cards[n] = c;
            n += 1;
        }
    }

    let mut best = HandRank::new(0, [0; 5]);
    let indices = [0usize, 1, 2, 3, 4, 5, 6];
    for i in 0..n {
        for j in (i + 1)..n {
            for k in (j + 1)..n {
                for l in (k + 1)..n {
                    for m in (l + 1)..n {
                        let five = [
                            cards[indices[i]],
                            cards[indices[j]],
                            cards[indices[k]],
                            cards[indices[l]],
                            cards[indices[m]],
                        ];
                        let rank = evaluate_five(&five);
                        if rank > best {
                            best = rank;
                        }
                    }
                }
            }
        }
    }
    best
}

fn card_rank(card: u8) -> u8 {
    card % 13
}

fn card_suit(card: u8) -> u8 {
    card / 13
}

fn evaluate_five(cards: &[u8; 5]) -> HandRank {
    let mut ranks = [0u8; 5];
    let mut suits = [0u8; 5];
    for (i, &c) in cards.iter().enumerate() {
        ranks[i] = card_rank(c);
        suits[i] = card_suit(c);
    }
    ranks.sort_unstable_by(|a, b| b.cmp(a));

    let mut rank_counts = [0u8; RANKS];
    let mut suit_counts = [0u8; SUITS];
    for &r in &ranks {
        rank_counts[r as usize] += 1;
    }
    for &s in &suits {
        suit_counts[s as usize] += 1;
    }

    let is_flush = suit_counts.iter().any(|&c| c >= 5);
    let straight_high = find_straight_high(&rank_counts);

    if is_flush && straight_high.is_some() {
        let high = straight_high.unwrap();
        if high == 12 {
            return HandRank::new(9, [12, 0, 0, 0, 0]); // Royal flush
        }
        return HandRank::new(8, [high, 0, 0, 0, 0]); // Straight flush
    }

    let mut quads = None;
    let mut trips = None;
    let mut pairs: [u8; 2] = [255, 255];
    let mut pair_count = 0usize;
    for r in (0..RANKS).rev() {
        let count = rank_counts[r];
        if count == 4 {
            quads = Some(r as u8);
        } else if count == 3 && trips.is_none() {
            trips = Some(r as u8);
        } else if count == 2 && pair_count < 2 {
            pairs[pair_count] = r as u8;
            pair_count += 1;
        }
    }

    if let Some(q) = quads {
        let kicker = (0..RANKS).rev().find(|&r| rank_counts[r] > 0 && r as u8 != q).unwrap() as u8;
        return HandRank::new(7, [q, kicker, 0, 0, 0]);
    }

    if let Some(t) = trips {
        if pair_count >= 1 {
            return HandRank::new(6, [t, pairs[0], 0, 0, 0]); // Full house
        }
    }

    if is_flush {
        return HandRank::new(5, ranks);
    }

    if let Some(high) = straight_high {
        return HandRank::new(4, [high, 0, 0, 0, 0]);
    }

    if let Some(t) = trips {
        let mut kickers = [0u8; 2];
        let mut ki = 0;
        for r in (0..RANKS).rev() {
            if rank_counts[r] > 0 && r as u8 != t && ki < 2 {
                kickers[ki] = r as u8;
                ki += 1;
            }
        }
        return HandRank::new(3, [t, kickers[0], kickers[1], 0, 0]);
    }

    if pair_count >= 2 {
        let high_pair = pairs[0].max(pairs[1]);
        let low_pair = pairs[0].min(pairs[1]);
        let kicker = (0..RANKS)
            .rev()
            .find(|&r| rank_counts[r] > 0 && r as u8 != high_pair && r as u8 != low_pair)
            .unwrap() as u8;
        return HandRank::new(2, [high_pair, low_pair, kicker, 0, 0]);
    }

    if pair_count == 1 {
        let pair = pairs[0];
        let mut kickers = [0u8; 3];
        let mut ki = 0;
        for r in (0..RANKS).rev() {
            if rank_counts[r] > 0 && r as u8 != pair && ki < 3 {
                kickers[ki] = r as u8;
                ki += 1;
            }
        }
        return HandRank::new(1, [pair, kickers[0], kickers[1], kickers[2], 0]);
    }

    HandRank::new(0, ranks)
}

fn find_straight_high(rank_counts: &[u8; RANKS]) -> Option<u8> {
    // Wheel: A-2-3-4-5
    if rank_counts[12] > 0
        && rank_counts[0] > 0
        && rank_counts[1] > 0
        && rank_counts[2] > 0
        && rank_counts[3] > 0
    {
        return Some(3); // 5-high straight
    }

    for high in (4..RANKS).rev() {
        let mut ok = true;
        for i in 0..5 {
            if rank_counts[high - 4 + i] == 0 {
                ok = false;
                break;
            }
        }
        if ok {
            return Some(high as u8);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    fn card(rank: u8, suit: u8) -> u8 {
        suit * 13 + rank
    }

    #[test]
    fn test_pair_beats_high_card() {
        let hole = [card(0, 0), card(0, 1)]; // pair of 2s
        let community = [card(12, 0), card(11, 1), card(10, 2), card(9, 3), card(8, 0)];
        let rank = evaluate_hand(hole, &community);
        assert_eq!(rank.category, 1);
    }
}
